Sming/system/include/m_printf.h

```diff
@@ -9,16 +9,66 @@ Descr: embedded very simple version of printf with float support
 #define _M_PRINTF_
 
 #include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+/** @brief callback type to output a string of data
+ *  @param param
+ *  @param str data to send
+ *  @param length number of characters to write
+ *  @retval number of characters written, which may be less than the requested size
+ *  @note data does not need to be nul terminated and may contain any 8-bit values including nul
+ */
+typedef size_t (*nputs_callback_t)(void* param, const char* str, size_t length);
+
+/** @brief set the character output routine
+ *  @param callback
+ *  @param param
+ */
+void m_setPuts(nputs_callback_t callback, void* param);
+
 int m_vsnprintf(char *buf, size_t maxLen, const char *fmt, va_list args);
 int m_snprintf(char* buf, int length, const char *fmt, ...);
 int m_printf(char const*, ...);
 int m_vprintf ( const char * format, va_list arg );
-void m_putc(char c);
+
+/** @brief output a single character
+ *  @param c
+ *  @retval size_t 1 on success, 0 if the character could not be output
+ */
+size_t m_putc(char c);
+
+/** @brief output a text string
+ *  @param str the text
+ *  @param length length of text, excluding nul terminator
+ *  @retval size_t number of characters actually output
+ *  @note nul terminator is optional
+ */
+size_t m_nputs(const char* str, size_t length);
+
+static inline size_t m_puts(const char* str)
+{
+	return m_nputs(str, strlen(str));
+}
+
+#ifdef __cplusplus
+#define M_PRINTHEX_BYTESPERLINE = 16
+#else
+#define M_PRINTHEX_BYTESPERLINE
+#endif
+
+/** @brief output a block of data in hex format
+ *  @param tag brief name to display with the data block
+ *  @param data
+ *  @param len
+ *  @param bytesPerLine If non-zero, data will be output in block separated by carriage return
+ *  @note intended for debugging
+ */
+void m_printHex(const char* tag, const void* data, size_t len, size_t bytesPerLine M_PRINTHEX_BYTESPERLINE);
```

Comment:
Just thinking out loud: Maybe we should get rid of the HexDump class and start using only this function to reduce the redundancy.

Reply:
I see that Hexdump outputs addresses, where as m_printHex doesn't. It could be added as another parameter. Also, if we restricted use of m_printHex to C++ code it would simplify the header, and we'd have the option of overloading it for more complex use cases.

```

/** @brief output a block of data in hex format
 *  @param tag brief name to display with the data block
 *  @param data
 *  @param len
 *  @param addr Prefix lines with addresses starting at the given value, use -1 if not required.
 *  @param bytesPerLine If non-zero, data will be output in block separated by carriage return
 *  @note intended for debugging
 */
void m_printHex(const char* tag, const void* data, size_t len, int addr = -1, size_t bytesPerLine = 16);
```

Reply:
Can add in this PR also the suggested changes?

Reply:
OK, I've revised m_printHex and tweaked it to get a more typical hex / ASCII output.
Tested using this snippet:

```
	static DEFINE_FSTR(test, "This is a some test data \1\2\3 Not all ASCII\0\0\0Anything, really\0Can I go home now?");
	LOAD_FSTR(data, test);
	m_printHex("TEST1", data, test.length());
	m_printHex("TEST2", data, test.length(), 0);
	m_printHex("TEST3", data, test.length(), 0xFFFFF9, 8);
	m_printHex(nullptr, data, test.length(), 0xFFFFF9, 0);
```